File: dataset/human/python/file_1054.py
from math import sqrt
def intersection(cx, cy, r, P1, P2):
    x1, y1 = P1; x2, y2 = P2

    xd = x2 - x1; yd = y2 - y1
    X = x1 - cx; Y = y1 - cy
    a = xd**2 + yd**2
    b = xd * X + yd * Y
    c = X**2 + Y**2 - r**2

    f0 = c; f1 = a+2*b+c
    if (f0 >= 0 and f1 <= 0) or (f0 <= 0 and f1 >= 0):
        return True
    return -a <= b <= 0 and b**2 - a*c >= 0 and (f0 >= 0 or f1 >= 0)

def solve():
    N = int(input())
    if N == 0:
        return False
    C = []
    for i in range(N):
        x, y, r = map(int, input().split())
        C.append((x, y, r))
    M = int(input())
    for i in range(M):
        tx, ty, sx, sy = map(int, input().split())
        for j in range(N):
            cx, cy, r = C[j]
            if intersection(cx, cy, r, (tx, ty), (sx, sy)):
                print("Safe")
                break
        else:
            print("Danger")
    return True
while solve():
    ...

File: dataset/human/java/file_1054.java
import java.util.Scanner;
public class Main
{
	public static void main(String[] args) 
	{
		Scanner in=new Scanner(System.in);
		for(;;)
		{
			int n=in.nextInt();
			if(n==0)
				return;
			double en[][]=new double[n][3];
			for(int i=0;i<n;i++)
				for(int j=0;j<3;j++)
					en[i][j]=in.nextDouble();
			int m=in.nextInt();
			for(int i=0;i<m;i++)
			{
				double tx=in.nextDouble(),ty=in.nextDouble();
				double sx=in.nextDouble(),sy=in.nextDouble();
				boolean judge=false;
				
				for(int j=0;j<n;j++)
				{
					double r=en[j][2];
					double dit=Math.hypot(tx-en[j][0],ty-en[j][1]);
					double dis=Math.hypot(sx-en[j][0],sy-en[j][1]);
					double d=lp_dis(en[j][0],en[j][1],tx,ty,sx,sy);
					if((dit>r||dis>r) && d<=r)
						judge=true;
				}
				System.out.println(judge ? "Safe":"Danger");
			}
		}
	}
	static double lp_dis(double px,double py,double x1,double y1,double x2,double y2)
	{
		double dx=x2-x1,dy=y2-y1;
		double a=dx*dx+dy*dy,b=dx*(x1-px)+dy*(y1-py);
		double t=-b/a;
		if(t<0) t=0;
		if(t>1) t=1;
		double tx=x1+dx*t;
		double ty=y1+dy*t;
		return Math.hypot(px-tx, py-ty);
	}
}
